enhancements/two-nodes-openshift/2no.md

```diff
@@ -0,0 +1,618 @@
+   * Pacemaker is [configured](//TODO mshitrit add link) to manage etcd/cri-o/kubelet, it will start/stop/restart those services using a script or an executable.
+   * Pacemaker does not understand what it is managing, and expects an executable or script that knows how to start/stop/monitor (and optionally promote/demote) the service.
+     1. Likely we would need to create one for etcd
+     2. For kubelet and cri-o we can likely use the existing systemd unit file
+2. Failure Scenarios:
+   * Implement detailed handling procedures for cold boots, network failures, node failures, kubelet failures, and etcd failures using the RHEL-HA stack.
+    [see examples](#failure-handling)
+3. Fencing Methods:
+   * We plan to use Baseboard Management Controller (BMC) as our primary fencing method, the premise of using BMC for fencing is that a node that is powered off, or was previously powered off and configured to be inert until quorum forms, is not in a position to cause corruption or diverging datasets.  Sending power-off (or reboot) commands to the peer’s BMC achieves this goal.
+
+
+### Workflow Description
+
+#### Cluster Creator Role:
+* The Cluster Creator will automatically install the 2NO (by using an installer), installation process will include the following steps:  
+  * Deploys a two-node OpenShift cluster
+  * Configures cluster membership and quorum using Corosync.
+  * Sets up Pacemaker for resource management and fencing.
+
+#### Application Administrator Role:
+* Receives cluster credentials.
+* Deploys applications within the two-node cluster environment.
+
+#### Failure Handling:
+
+1. Cold Boot
+   1. One node (Node1) boots
+   2. Node1 does have “corosync quorum” (C-quorum)  (requires forming a membership with it’s peer)
+   3. Node1 does not start etcd or kubelet, remains inert waiting for Node2
+   4. Peer (Node2) boots
+   5. Corosync membership containing both nodes forms
+   6. Pacemaker “starts” etcd on both nodes
+      * Detail, this could be a “soft”-start which allows us to determine which node has the most recent dataset.
+   7. Pacemaker “promotes” etcd on whichever node has the most recent dataset
+   8. Pacemaker “promotes” etcd on the peer once it has caught up
+   9. Pacemaker starts kubelet on both nodes
+   10. Fully functional cluster
+2. Network Failure
+   1. Corosync on both nodes detects separation
+   2. Etcd loses internal quorum (E-quorum) and goes read-only
+   3. Both sides retain C-quorum and initiate fencing of the other side. There is a different delay between the two nodes for executing the fencing operation to avoid both fencing operations to succeed in parallel and thus shutting down the system completely.
+   4. One side wins, pre-configured as Node1
+   5. Pacemaker on Node1 forces E-quorum (etcd promotion event)
```

Comment:
voting

Reply:
When etcd loses quorum, and you have two halves, both are in read only states.

Have you spoken to the etcd folks/documented what is actually involved in this step?

It's been a while since I've had to do this, so things may have changed, but my understanding is that the only way to recover from this is to either reconnect them, or, force a "new cluster" using the old state, which removes all members and starts the cluster again. On the other member you'd also need to remove state and have it come us as a fresh etcd, resync across and then promote again.

"forcing quorum" makes this sound easy, but I think the document currently underplays exactly what's involved in this process

Reply:
Extensive consultation with the etc team :-)
The updated proposal should make things clearer, or do you think we need the exact commands?

Reply:
Maybe not exact commands, but a quick delve into the process and risks of forcing etcd quorum might be useful context?

Reply:
IIUC the process of forcing e-quorum boils down to triggering  the `MemberPromote` [API](https://etcd.io/docs/v3.5/dev-guide/api_reference_v3/).

@tjungblu please let me know if I got it right and if there are particular risks we should consider.

Reply:
@JoelSpeed, you are right, it sounds easier that what it is.
We are considering and discussing the cases and states together with the Etcd team.

> quick delve into the process

It is not quick :smile:, but it boils down mostly in the two ways a node can go down

* gracefully
* non-gracefully

In the first, the shutting down node can remove itself from the member list, which let the surviving node continue standalone without rebooting into a new cluster.

In the second, the surviving node must force a new cluster, instead.

The restarting node will decide what to do based on the status of the other node (running, starting, or stopped) and on data about Cluster ID and Revision from _both_ nodes.

We didn't complete the picture of the cases, but I wonder what level of detail we want to report in this enhancement

Reply:
Personally, I'd be keen to see the process documented fairly thoroughly, as it's rather key to the project working and being safe

Reply:
updated 156c11aceb22bf1b7fcfa6e82d0cdcbe49aec9b9